### ISSUE TITLE
Distributor crashes with ha tracker configured for memberlist. Does ha_tracker support memberlist kvstore ?

### ISSUE THREAD

**Issue by paulsinai** [cortexproject/cortex#2929]
Hi,

I'm trying to deploy a distributed cortex on GCP with a deployment or statefulset for each components. I started by using ETCD but I faced very quickly issues in term of maintenance with a DB size growing up to 2,2 GB in 2 days ! Then I tried to switch on memberlist in order to reduce the dependency to ETCD. It works well for all components even the distributor but not for the ha tracker. 

1. I start the ingesters -> Works 
2. I start any components (usually the distributor) -> The ingesters works but the distributor fails

Version: 1.2.0
Environment: GKE 1.16.11

**Logs from Distributor**
```
"level=debug ts=2020-07-24T23:19:51.120546821Z caller=api.go:94 msg="api: registering route" methods= path=/ring auth=false"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x26bdffc]
goroutine 1 [running]:
github.com/cortexproject/cortex/pkg/ring/kv.createClient(0xc000133db0, 0xa, 0xc000133dc0, 0xb, 0x364518e, 0xe, 0x0, 0x0, 0x4a817c800, 0x0, ...)
	/go/src/github.com/cortexproject/cortex/pkg/ring/kv/client.go:130 +0x54c
github.com/cortexproject/cortex/pkg/ring/kv.NewClient(0xc000133db0, 0xa, 0xc000133dc0, 0xb, 0x364518e, 0xe, 0x0, 0x0, 0x4a817c800, 0x0, ...)
	/go/src/github.com/cortexproject/cortex/pkg/ring/kv/client.go:107 +0x117
github.com/cortexproject/cortex/pkg/distributor.newClusterTracker(0x1, 0x37e11d600, 0x12a05f200, 0x6fc23ac00, 0xc000133db0, 0xa, 0xc000133dc0, 0xb, 0x364518e, 0xe, ...)
	/go/src/github.com/cortexproject/cortex/pkg/distributor/ha_tracker.go:160 +0x1d1
github.com/cortexproject/cortex/pkg/distributor.New(0x37e11d600, 0x1, 0x77359400, 0x1, 0x37e11d600, 0x12a05f200, 0x6fc23ac00, 0xc000133db0, 0xa, 0xc000133dc0, ...)
	/go/src/github.com/cortexproject/cortex/pkg/distributor/distributor.go:187 +0xe7
github.com/cortexproject/cortex/pkg/cortex.(*Cortex).initDistributor(0xc000a3d000, 0x0, 0x0, 0x0, 0xc00047c270)
	/go/src/github.com/cortexproject/cortex/pkg/cortex/modules.go:154 +0x1d9
github.com/cortexproject/cortex/pkg/util/modules.(*Manager).InitModuleServices(0xc000388078, 0x7ffc93c3b0ee, 0x5, 0x3d98c00, 0xc000881b60, 0x3e109e0)
	/go/src/github.com/cortexproject/cortex/pkg/util/modules/modules.go:72 +0x26f
github.com/cortexproject/cortex/pkg/cortex.(*Cortex).Run(0xc000a3d000, 0xc000694c40, 0x4)
	/go/src/github.com/cortexproject/cortex/pkg/cortex/cortex.go:294 +0x5b
main.main()
	/go/src/github.com/cortexproject/cortex/cmd/cortex/main.go:135 +0xa18
```

**My conf (Truncated):**

```  auth_enabled: false
    server:
      http_listen_address: ""
      http_listen_port: 80

      grpc_listen_address: ""
      grpc_listen_port: 9095

      log_level: "debug"

      grpc_server_max_recv_msg_size: 104857600
      grpc_server_max_send_msg_size: 104857600
      grpc_server_max_concurrent_streams: 1000

    memberlist:
      randomize_node_name: true
      join_members:
        - ingester-cortex-headless.default.svc.cluster.local:7946

      min_join_backoff: 1s
      max_join_backoff: 1m
      max_join_retries: 10
      abort_if_cluster_join_fails: true
      bind_port: 7946

    distributor:
      pool:
        client_cleanup_period: 15s
        health_check_ingesters: true

      ha_tracker:
        enable_ha_tracker: true
        ha_tracker_update_timeout: 15s
        ha_tracker_update_timeout_jitter_max: 5s
        ha_tracker_failover_timeout: 30s

        kvstore:
          store: "memberlist"
          prefix: "ha-tracker/"
          
      remote_timeout: 2s
      extra_queue_delay: 0s
      shard_by_all_labels: true

      ring:
        kvstore:
          store: "memberlist"
          prefix: "collectors/"
        heartbeat_period: 5s
        heartbeat_timeout: 1m
```

I successfully launched the cluster in memberlist when I left the ha tracker on ETCD.

Should I remove the kvstore configuration from the ha_tracker config ? In this case, the error should be handled and the documentation updated. Thanks for your help.

**Comment by pstibrany** [comment]
No, using memberlist for HA is not supported. Memberlist-based KV store propagates updates using gossip, which is very slow for HA purposes: result is that different distributors may see different Prometheus server as elected HA replica, which is definitely not desirable.

(I wouldn’t expect a panic though, that’s something to be checked and fixed)

**Comment by paulsinai** [comment]
Thanks for the answer. I suggest to start by modifying the documentation because memberlist is mentioned as supported. Could you confirm Cortex HA Tracker does support Consul, ETCD, inmemory and multi only (not memberlist) ?
I can fix the doc and make a PR for this if you want.

**Comment by pstibrany** [comment]
> Thanks for the answer. I suggest to start by modifying the documentation because memberlist is mentioned as supported. Could you confirm Cortex HA Tracker does support Consul, ETCD, inmemory and multi only (not memberlist) ?
> I can fix the doc and make a PR for this if you want.

To make sure HA tracker works correctly, it should use Consul or Etcd. "inmemory" is inside single process only, and only works if there is single distributor.

"multi" is used for migrating between KV stores without downtime (eg. from Consul to Etcd or vice-versa), but it isn't designed to work with HA properly.

**Comment by pracucci** [comment]
> I suggest to start by modifying the documentation

@paulbellamy Thanks for raising this lack in the doc. Could you open a PR to improve our doc, please?

**Comment by pracucci** [comment]
Closed by #3077.
